### ISSUE TITLE
On some devices, a Security Risk warning is displayed for api, relay, vr subdomains

### ISSUE THREAD

**Issue by parseccentric** [gelic-idealab/komodo-portal#28]
I might have ventured out pretty far on my own. I just wanted to do a quick test and then ended up tumbling down the road of learning.  Here's what I've learned so far: 

I ran our domain through some SSL checkers online: 

**Liquid Web SSL Checker**

— https://sslcheck.liquidweb.com/?domain=relay.komodo-dev.library.illinois.edu
Valid but not trusted: dash, api, relay, vr

**Qualys SSL Server Test**
— https://www.ssllabs.com/ssltest/analyze.html?d=komodo-dev.library.illinois.edu

The website says this is why we are not valid but not trusted: 

> Extra download        R3
> Fingerprint SHA256: 730c1bdcd85f57ce5dc0bba733e5f1ba5a925b2a771d640a26f7a454224dad3b
> Pin SHA256: jQJTbIh0grw0/1TkHSumWb+Fs0Ggogr621gT3PvPKG0=
> RSA 2048 bits (e 65537) / SHA256withRSA

A StackOverflow answer sheds some light: 

> You have configured your server to only send the certificate to browsers. For most desktop browsers, this is fine because they already contain a whole lot of intermediate and root CA details, so they can construct the chain of trust easily. For most mobile browsers you generally need to provide the entire certificate chain, i.e. your own certificate, that of the issuing CA and any intermediates that might exist between that and the ultimate root CA. The mobile device will likely have the root CA details only in this scenario.
> 
> For you specific cert, you can read this Comodo helpdesk article: Knowledgebase : Comodo Certification Authority > Certificates > SSL > Certificate Installation
> 
> -- “ThatGraemeGuy’s Answer to What is wrong with my SSL trust chain?,” Server Fault (StackExchange), Sep. 12, 2014. https://serverfault.com/questions/628062/what-is-wrong-with-my-ssl-trust-chain/628063#628063 (accessed Apr. 14, 2021).

**According to [Comodo Support > Certificate Installation: Apache & mod_ssl](https://support.comodo.com/index.php?/Knowledgebase/Article/View/637/37/certificate-installation-apache--mod_ssl), the steps needed for an Apache server are…** 

- retrieve the *.crt and *.ca-bundle files
- move the SSH Private key to the right directory.
- Move the *.crt and *.ca-bundle files to the right directory. 
- Use the Apache configuration file to find the SSL configuration file.
- In the SSL config file, set…
    - SSLEngine
    - SSLCertificateKeyFile: SSH Private key
    - SSLCertificateFile: *.crt
    - SSLCertificateChainFile: *.ca-bundle
- Save config file and restart. 

(A similar article for Nginx led to a 404, but we can generalize the Apache instructions to an Nginx server, and our traefik configuration.)

**Following the instructions on [Nginx Docs > ... > Configuring HTTPS Servers: SSL Certificate Chains](https://nginx.org/en/docs/http/configuring_https_servers.html#chains)**

It seems we need a *bundle.crt or *chained.crt file. 

> Browsers usually store intermediate certificates which they receive and which are signed by trusted authorities, so actively used browsers may already have the required intermediate certificates and may not complain about a certificate sent without a chained bundle. To ensure the server sends the complete certificate chain, the `openssl` command-line utility may be used, for example:
    openssl s_client -connect komodo-dev.library.illinois.edu:443

Running that line resulted in 

```
    Certificate chain
     0 s:CN = komodo-dev.library.illinois.edu
       i:C = US, O = Let's Encrypt, CN = R3
```

and nothing else for the chain.

> If a certificate bundle has not been added, only the server certificate #0 will be shown.

Seems like we need to follow this: 

> The server certificate must appear before the chained certificates in the combined file:
    $ cat www.example.com.crt bundle.crt > www.example.com.chained.crt
> The resulting file should be used in the ssl_certificate directive:

```
    server {
        listen              443 ssl;
        server_name         www.example.com;
        ssl_certificate     www.example.com.chained.crt;
        ssl_certificate_key www.example.com.key;
        ...
    }
```

I think this helps with the nginx config, but what about traefik? It needs to know about the certs, too, right? 

This do-it-yourself solution looks relevant: 

> dtomcej: I think that it is not an unfair expectation for users to concatenate the user, intermediate and CA certs if required into a single PEM.
> 
> nmegin: I solved the problem with a little script ran when the container is started.
> It detects if there is a CA file and cert files and concatenates them.
> 
> -- traefik/traefik/issues: [Add parameter to configure TLS entrypoints with ca-bundle file #984](https://github.com/traefik/traefik/issues/984)

Further reading: 

- https://support.dnsimple.com/articles/what-is-ssl-certificate-chain/
- There is a programmatic solution written in Go: zakjan/cert-chain-resolver
- And an online certificate concatenator, but that seems rather sketchy to me: https://certificatechain.io/

tl;dr: It looks like we need to add a *.ca-bundle to our configuration. Further, it looks like we may need to concatenate the chain ourselves.

**Comment by parseccentric** [comment]
Resolved! From now on, we need to use the provided `*-fullchain.crt file`, which is already concatenated. 

Revisit the Qualys SSL Server Test and https://www.ssllabs.com/ssltest/analyze.html?d=komodo-dev.library.illinois.edu . Make sure to clear the cache. 

The results should say "A" and Additional Certificates > Chain issues should say "none".
